### ISSUE TITLE
Generating collection methods to RootModel

### ISSUE THREAD

**Issue by alkshmir** [koxudaxi/datamodel-code-generator#1830]
**Is your feature request related to a problem? Please describe.**
When I convert an array like following with `datamodel-codegen --input openapi.yaml --output model.py --output-model-type pydantic_v2.BaseModel`,
```yaml
openapi: 3.0.3

components:
  schemas:
    MyItem:
      type: object
      properties:
        foo:
          type: number
        bar:
          type: string

    MyArray:
      type: array
      items:
        $ref: "#/components/schemas/MyItem"
```

`MyArray` cannot be used like a list because it lacks methods that collections should have.

For example, the following code prints `('root', [MyItem(foo=123.0, bar='text')])` instead of `MyItem(foo=123.0, bar='text')`.

```python
from model import MyArray, MyItem

arr = MyArray([MyItem(foo=123, bar="text")])

for item in arr:
    print(item)
```

**Describe the solution you'd like**
It would be nice to generate the methods that collections should have, e.g., `__getitem__`, `__iter__`, or `__len__`:

```python
class MyItem(BaseModel):
    foo: Optional[float] = None
    bar: Optional[str] = None


class MyArray(RootModel[List[MyItem]]):
    root: List[MyItem]

    def __getitem__(self, index):
        return self.root[index]

    def __iter__(self):
        return iter(self.root)

    def __len__(self):
         return len(self.root)
```

This would be a common use case so it would be great that some CLI option enables this feature.

**Describe alternatives you've considered**
Writing a collection methods manually on every generated `RootModel`, or pydantic might have some other pretty solutions.

**Additional context**
Other collection methods like `__setitem__`, `__delitem__`, `__contains__`, or some other methods list-like collections have
(`append`, `extend`, `insert`, `remove`, `pop`, `clear`, or `index`) might be better to generated as well.

Sorry if this request was a duplicate.
Thank you very much for the great tool.

**Comment by rafalkrupinski** [comment]
```python
for item in arr.root:
    print(item)
```

**Comment by alkshmir** [comment]
Thanks for your help.
Accessing `root` directly can be a solution but I'm just wondering how to use the model in the same way as `List[MyItem]` not to depend on the internal structure.

**Comment by rafalkrupinski** [comment]
Technically `root` is a part of pydantics public API, like `model_extra` and other properties, but I understand it may be an inconvenience.

**Comment by alkshmir** [comment]
I did not check the discussion but this may help me (but it seems they are talking about pydantic v1 and idk it will work for v2 as well): https://github.com/koxudaxi/datamodel-code-generator/discussions/640

**Comment by koxudaxi** [comment]
@rafalkrupinski 
Thank you for your help.

@alkshmir 

As your link shows, this would be a good solution to use a custom template and run the CLI to generate your own root model.
What do you think?

**Comment by JakubKahovec** [comment]
I really like @alkshmir idea adding ability to generate the collection methods, allowing to work with the model like if it was a collection i.e `list` or `dict`. Having to access the `root` field goes almost against the idea of the RootModel which somehow tries to hide the underlying structure.

**Comment by rafalkrupinski** [comment]
My generator generates model just for the item and uses `list[item]`.
It's not like JSON can represent anything that's both object and array at the same time.

**Comment by JakubKahovec** [comment]
I didn't mean it'd be both. I meant that if the RootModel would be `dict` it'd generate the magic methods for `dict` so it'd be possible to work with the class like with a dict and it'd be `list` it'd generate method for list.

**Comment by rafalkrupinski** [comment]
I mean, why generate a class for `type: array` in the first place?
This request asks to make code more complex when it could be simpler.

**Comment by JakubKahovec** [comment]
Well, that’s the main point of the [RootModel](https://docs.pydantic.dev/latest/concepts/models/#rootmodel-and-custom-root-types) class. 
It can be for better type safety or for validation purposes.

**Comment by evilhamsterman** [comment]
I agree with @JakubKahovec the point of RootModel is two fold. It helps with hinting by and validation by enforcing at execution `Users[User]` where an annotated type is just a hint and not enforced.

**Comment by koxudaxi** [comment]
This behavior comes from Pydantic v2's RootModel design itself. Iterating over a RootModel iterates over model fields rather than the root value. The recommended way is to access .root directly:
```python
  for item in arr.root:
      print(item)
```
If you prefer not to use .root, you can use --use-type-alias to generate a TypeAliasType instead (though this removes runtime validation), or use --custom-template-dir with a modified template that includes collection methods.
